src/mlpack/methods/ann/layer/adaptive_max_pooling_impl.hpp

```diff
@@ -43,6 +43,59 @@ AdaptiveMaxPooling<InputDataType, OutputDataType>::AdaptiveMaxPooling(
   poolingLayer = ann::MaxPooling<>(0, 0);
 }
 
+template<typename InputDataType, typename OutputDataType>
+AdaptiveMaxPooling<InputDataType, OutputDataType>::AdaptiveMaxPooling(
+  const AdaptiveMaxPooling& layer) :
+  outputWidth(layer.outputWidth),
+  outputHeight(layer.outputHeight),
+  reset(layer.reset),
+  poolingLayer(layer.poolingLayer)
+{
+  // Nothing to do here.
+}
+
+template <typename InputDataType, typename OutputDataType>
+AdaptiveMaxPooling<InputDataType, OutputDataType>::AdaptiveMaxPooling(
+    AdaptiveMaxPooling&& layer) :
+    outputWidth(std::move(layer.outputWidth)),
+    outputHeight(std::move(layer.outputHeight)),
+    reset(std::move(layer.reset)),
+    poolingLayer(std::move(layer.poolingLayer))
+{
+   // Nothing to do here.
+}
+    
+template <typename InputDataType, typename OutputDataType>
+AdaptiveMaxPooling<InputDataType, OutputDataType>&
+AdaptiveMaxPooling<InputDataType, OutputDataType>::
+operator=(const AdaptiveMaxPooling& layer) 
+{
+  if (this != &layer)
+  {
+    outputWidth = layer.outputWidth;
+    outputHeight = layer.outputHeight;
+    reset = layer.reset;
+    poolingLayer = layer.poolingLayer;
+  }
+  return *this;
+}
+    
+template <typename InputDataType, typename OutputDataType>
+AdaptiveMaxPooling<InputDataType, OutputDataType>&
+AdaptiveMaxPooling<InputDataType, OutputDataType>::
+operator=(AdaptiveMaxPooling&& layer) 
+{
+  if (this != &layer)
+  {
+    outputWidth = std::move(layer.outputWidth);
+    outputHeight = std::move(layer.outputHeight);
+    reset = false;
```

Comment:
I can't test this right now, but I think we have to reset the layer everytime, can you set `reset = true;` and see if that works, we might have to do the same for `operator=` as well.

Reply:
reset was set to false initially.. but yes it gave some warnings regarding the same as well as for the poolingLayer while building.

Reply:
Interesting, based on the test error some part of the network is not copied correctly. But unless you already have a running gdb setup to step through the code, I could test this locally.

Reply:
I will setup one.  I don't have now.

Reply:
```
-------------------------------------------------------------------------------
CheckCopyMovingAdaptiveMaxNetworkTest
-------------------------------------------------------------------------------
/home/naba/dev/mlpack/src/mlpack/tests/convolutional_network_test.cpp:443
...............................................................................

/home/naba/dev/mlpack/src/mlpack/tests/convolutional_network_test.cpp:443: FAILED:
due to unexpected exception with message:
  FFN<>::Train(): the first layer of the network expects 9216 elements, but the
  input has 784 dimensions!

-------------------------------------------------------------------------------
CheckCopyMovingAdaptiveMeanNetworkTest
-------------------------------------------------------------------------------
/home/naba/dev/mlpack/src/mlpack/tests/convolutional_network_test.cpp:531
...............................................................................

/home/naba/dev/mlpack/src/mlpack/tests/convolutional_network_test.cpp:531: FAILED:
due to unexpected exception with message:
  FFN<>::Train(): the first layer of the network expects 9216 elements, but the
  input has 784 dimensions!

--info: At Most 1 required but received 2
The following argument was not expected: --unknown
```

Reply:
I get this error

Reply:
If you wanted to merge this PR, we'd still need to work out this issue.  However, I'm not sure if it's worth it, because this code is getting completely refactored in #2777 and will probably end up with a different implementation for the copy/move constructor.

Reply:
I don't get this issue now. It is a slate one.

Reply:
I had hoped that the work coming out of the issue would still be useful, but it's only become clear in the past day or so that the copy/move constructors/operators look pretty different in #2777.  I'll post a comment there later today.  Definitely these things are still needed!  They are just implemented somewhat differently after the refactoring.  Sorry if that is a little frustrating given the work you've already put in---there are lots of moving pieces and it's hard to make them all sync up. :(

Reply:
yea, no worries! Ah, I will look into #2777 and find something there to help. Always happy to learn and help!! :)